lib/spack/spack/directives.py

```diff
@@ -189,15 +196,44 @@ def _decorator(decorated_function):
 
             @functools.wraps(decorated_function)
             def _wrapper(*args, **kwargs):
+                # If any of the arguments are executors returned by a
+                # directive passed as an argument, don't execute them
+                # lazily. Instead, let the called directive handle them.
+                # This allows nested directive calls in packages.  The
+                # caller can return the directive if it should be queued.
+                def remove_directives(arg):
+                    directives = DirectiveMetaMixin._directives_to_be_executed
+                    if isinstance(arg, (list, tuple)):
+                        # Descend into args that are lists or tuples
+                        for a in arg:
+                            remove_directives(a)
+                    else:
+                        # Remove directives args from the exec queue
+                        remove = next(
+                            (d for d in directives if d is arg), None)
+                        if remove is not None:
+                            directives.remove(remove)
+
+                # Nasty, but it's the best way I can think of to avoid
+                # side effects if directive results are passed as args
+                remove_directives(args)
```

Comment:
I'm trying to wrap my head around the directives logic and your updates to it. Could you provide an example of what goes wrong when this isn't done?

I feel like there's got to be a way to avoid undoing undesired modifications, but I admit I don't get what's going on here yet.

Reply:
Sure.  Say you do this:

```python
class Foo(Package):
    depends_on('bar', patches=[
        patch('bar-bugfix.patch'),
        patch('https://example.com/bar-feature.patch', sha256='abc123...'))
```

If you do not remove the results of the two `patch` calls from the queue here, they'll be lazily executed along with all the other directives in `DirectiveMetaMixin.__init__` (the metaclass initializer), because the `@directive` decorator adds them to `DirectiveMetaMixin._directives_to_be_executed`.  The point of this is to *only* pass the result of `patch()` to the `depends_on()` directive, and let that directive handle it.

Basically the point of all the extra directive machinery is so that only directives at the class definition level get executed.  The others are either nested (e.g., like `patch()` in `_depends_on`, which is only executed for its side effects) or they're passed as parameters, and we let the callee directive decide what to do with their results.  The callee (`depends_on`) could execute the `_execute_patch` function objects itself, or it could conceivably enqueue them.